File: aoc/2021/aoc202102.hs
move :: (Int,Int) -> [String] -> (Int,Int)
move (x,y) ("forward":n:rest) = move (x+read n,y) rest
move (x,y) ("down":n:rest) = move (x,y+read n) rest
move (x,y) ("up":n:rest) = move (x,y-read n) rest
move (x,y) _ = (x,y)

testData :: String
testData = "forward 5\ndown 5\nforward 8\nup 3\ndown 8\nforward 2\n"

test :: ()
test
  | (move (0,0) . words) testData /= (15,10) = error "a"
  | otherwise = ()

part1 :: IO Int
part1 = fmap (uncurry (*) . move (0,0) . words) $ readFile "input/02.txt"

move2 :: ((Int,Int),Int) -> [String] -> ((Int,Int),Int)
move2 ((x,y),aim) ("forward":n:rest) = move2 ((x+read n,y+aim*read n),aim) rest
move2 ((x,y),aim) ("down":n:rest) = move2 ((x,y),aim+read n) rest
move2 ((x,y),aim) ("up":n:rest) = move2 ((x,y),aim-read n) rest
move2 ((x,y),aim) _ = ((x,y),aim)

test2 :: ()
test2
  | (move2 ((0,0),0) . words) testData /= ((15,60),10) = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap (uncurry (*) . fst . move2 ((0,0),0) . words) $ readFile "input/02.txt"

File: aoc/2025/aoc202102.go
package main

func init() {
	Register(&aoc202102{
		AOC: AOC{
			Day:           2,
			InputFilename: "../2021/input/02.txt",
			Tests: []Test{
				Test{
					Input: `forward 5
down 5
forward 8
up 3
down 8
forward 2
`,
					Part1: "150",
					Part2: "900",
				},
			},
		},
	})
}

type aoc202102 struct {
	AOC
}

func (aoc *aoc202102) Part1(input *Input) string {
	x, y := 0, 0
	for {
		dir, ok := input.Word()
		if !ok {
			break
		}
		n, _ := input.Int()
		switch dir {
		case "forward":
			x += n
		case "up":
			y -= n
		case "down":
			y += n
		}
	}
	return IntResult(x * y)
}

func (aoc *aoc202102) Part2(input *Input) string {
	x, y, aim := 0, 0, 0
	for {
		dir, ok := input.Word()
		if !ok {
			break
		}
		n, _ := input.Int()
		switch dir {
		case "forward":
			x += n
			y += n * aim
		case "up":
			aim -= n
		case "down":
			aim += n
		}
	}
	return IntResult(x * y)
}
